### ISSUE TITLE
metrics values don't match between custom metric function and model.fit callbacks

### ISSUE THREAD

**Issue by asterisk37n** [tensorflow/tensorflow#47366]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Google Colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): Google Colab default installed
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.10
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**  
returned metrics value from custom metrics functions and callback printed metrics from model.fit doesn't match.

**Describe the expected behavior**  
returned metrics value from custom metrics functions and callback printed metrics from model.fit match.

**Standalone code to reproduce the issue**  
[Google Colab gist](https://gist.github.com/asterisk37n/9e8c492fc7f01a9f37f9d5fb33ee3406)

**Other info / logs**  
I opened a quesiton on stackoverflow.  
[https://stackoverflow.com/questions/66311611/when-and-how-keras-calculate-metrics-for-each-batch-of-samples](https://stackoverflow.com/questions/66311611/when-and-how-keras-calculate-metrics-for-each-batch-of-samples)

**Comment by ariG23498** [comment]
Hey @asterisk37n 
I tried giving your script a run, modifying a little bit
```python
history = model.fit(x=x, y=y, batch_size=2, epochs=1, verbose=2)
```
Here I want to output only one line for each epoch and not the progress bar, hence `verbose=2`
Upon doing this, I do in fact get results that are aligned with the `history.history`

My guesses are to dive deep into the output of `model.fit` and check whether the progress bar messes up with the results that we see. Hope this helps 😄

**Comment by evgeniya-egupova** [comment]
Hi all. Here are the issues describing the same problem: #46713 #47099. Hope this helps.

**Comment by amahendrakar** [comment]
@asterisk37n,
On running the Colab notebook you have linked, I see that the `loss` and `my_metric_fn` values are identical. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/caa1b8fc1c251ef54a4fbc8395ec0003/47366.ipynb#scrollTo=EZK_OjvBY4Pj). Thanks!

**Comment by asterisk37n** [comment]
@amahendrakar 
Thank you for running. I'm referring to a different thing. In my log, you can see 0.1429. Where has this come from?
```
2/2 [==============================] - 0s 7ms/step - loss: 0.1429 - my_metric_fn: 0.1429
```
```my_metric_fn```'s value for the second batch? No, it is 0.0450076237.
Average of ```my_metric_fn``` among the first and second batch? No, it is (0.191785976 +  0.0450076237)/2 = 0.11839679985.

**Comment by asterisk37n** [comment]
@ariG23498 Thanks for verbose param info.
@evgeniya-egupova  Thank you. yeah, very similar to https://github.com/tensorflow/tensorflow/issues/47099

**Comment by ricvo** [comment]
Could that be a running average with some exp decay parameter? i.e. not a simple average.
Might be linked to https://github.com/tensorflow/tensorflow/issues/39448

**Comment by ricvo** [comment]
related issue https://github.com/tensorflow/tensorflow/issues/47099 has been closed, could you check if the present issue is solved too on the new commit by @rchao?

**Comment by rchao** [comment]
@ricvo very likely so. If you find that the original issue is still not fixed after using a nightly build (that has the fix), please comment. Thanks!

**Comment by jvishnuvardhan** [comment]
@asterisk37n As mentioned by @rchao and @ricvo, this was resolved with recent `tf-nightly`. Please check the [gist here]().

Old output (with `TF2.4.1`)

```
2.4.1
TensorShape([2, 1]) TensorShape([2, 1]) [0.216433451 0.167138502] 0.191785976
1/2 [==============>...............] - ETA: 0s - loss: 0.1918 - my_metric_fn: 0.1918
TensorShape([2, 1]) TensorShape([2, 1]) [0.0477369316 0.0422783121] 0.0450076237
2/2 [==============================] - 0s 12ms/step - loss: 0.1429 - my_metric_fn: 0.1429
{'loss': [0.1183968037366867], 'my_metric_fn': [0.1183968037366867]}
```

New output (with `tf-nightly`)


```
Epoch TensorShape([2, 1]) TensorShape([2, 1]) [0.216433451 0.167138502] 0.191785976
1/2 [==============>...............] - ETA: 0s - loss: 0.1918 - my_metric_fn: 0.1918
Epoch TensorShape([2, 1]) TensorShape([2, 1]) [0.0477369316 0.0422783121] 0.0450076237
2/2 [==============================] - 0s 15ms/step - loss: 0.1184 - my_metric_fn: 0.1184
{'loss': [0.1183968037366867], 'my_metric_fn': [0.1183968037366867]}
```

Please verify once and close the issue. Thanks!

**Comment by asterisk37n** [comment]
@jvishnuvardhan Thank you. I checked your gist and it works perfectly as expected! Thank you for contribution.

**Comment by jvishnuvardhan** [comment]
Thanks for confirming. Closing the issue as this was resolved.
